Users/binah/access-control/access-control.cs

```diff
@@ -0,0 +1,320 @@
+            if (DebugLogging) Log.Write("Say goodbye to " + agent.AgentInfo.Name);
+        }
+        catch (NullReferenceException nre) { if (DebugLogging) Log.Write("Bannish", nre.Message); } // User Gone.
+        catch (System.Exception e) { if (DebugLogging) Log.Write("Bannish", e.ToString()); }
+    }
+
+    private void GrantEntry(AgentPrivate agent)
+    {
+        //if on ban list remove them
+        if(IsBanned(agent))
+        {
+            RemoveUserBan(agent);
+        }
+        //setAccess
+        SetAccess(agent);
+    }
+
+    private void AskForEntry(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write("Checking Access");
+        string message = "Allow " + agent.AgentInfo.Name + " to enter?";
+        Visitor admin = Visitors[Admins[0]];
+        ModalDialog modalDialog = admin.Agent.Client.UI.ModalDialog;
+        modalDialog.Show(message, "Yes", "No", (opc) =>
+        {
+            if(modalDialog.Response == "Yes")
+            {
+                if (DebugLogging) Log.Write("Entry was granted.");
+                GrantEntry(agent);
+            } else
+            {
+                if (DebugLogging) Log.Write("Entry was denied.");
+                BanUser(agent);
+            }
+        });
+    }
+
+    private void BanUser(AgentPrivate agent)
+    {
+        Banned.Add(agent.AgentInfo.Handle.ToLower());
+        if (DebugLogging) Log.Write(agent.AgentInfo.Name + " has been added to banlist");
+        Bannish(agent);
+    }
+
+    private void RemoveUserBan(AgentPrivate agent)
+    {
+        Banned.RemoveAll(a => a == agent.AgentInfo.Handle.ToLower());
+        if (DebugLogging) Log.Write(agent.AgentInfo.Name + " has been removed from the banlist");
+    }
+
+    private void ToggleDoors(ChatData data)
+    {
+        if (DebugLogging) Log.Write("Toggling doors " + data);
+        if (data.Message != ToggleDoorCommand)
+        {
+            return;
+        }
+
+        if (DebugLogging) Log.Write("ToggleDoors: " + !DoorsOpen);
+        DoorsOpen = !DoorsOpen;
+    }
+
+    private void ShowLogCommand(ChatData data)
+    {
+        // Checking the message is the fastest thing we could do here. Discard anything that isn't the command we are looking for.
+        if (data.Message != VisitorListCommand)
+        {
+            return;
+        }
+
+        AgentPrivate agent = ScenePrivate.FindAgent(data.SourceId);
+        if (agent == null)
+        {
+            if (DebugLogging) Log.Write("Possible race condition and they already logged off.");
+            return;
+        }
+
+        if (IsAdmin(agent) || agent.AgentInfo.AvatarUuid == ScenePrivate.SceneInfo.AvatarUuid)
+        {
+            ShowLog(agent);
+        }
+    }
+
+    private void OnBanCommand(ChatData data)
+    {
+        if (DebugLogging) Log.Write("Ban Command triggered.");
+
+        string[] chatWords = data.Message.Split(' ');
+
+        if (String.IsNullOrEmpty(chatWords[0]) || String.IsNullOrEmpty(chatWords[1]))
```

Comment:
there might be no spaces in the message, so `chatWords[1]` would throw an exception that case